src/zql/ivm/graph/difference-stream.ts

```diff
@@ -43,48 +49,83 @@ import {QueueEntry} from './queue.js';
  */
 // T extends object: I believe in the context of ZQL we only deal with object.
 export class DifferenceStream<T extends object> {
-  readonly #upstreamWriter;
+  readonly #downstreams: Set<Listener<T>> = new Set();
+  #upstream: Operator | undefined;
+  readonly #requestors = new Set<Listener<T>>();
+
+  addDownstream(listener: Listener<T>) {
+    this.#downstreams.add(listener);
+  }
 
-  constructor(upstreamWriter?: DifferenceStreamWriter<T> | undefined) {
-    this.#upstreamWriter = upstreamWriter ?? new DifferenceStreamWriter<T>();
+  setUpstream(operator: Operator) {
+    invariant(this.#upstream === undefined, 'upstream already set');
+    this.#upstream = operator;
+    return this;
+  }
+
+  newData(
+    version: Version,
+    data: Iterable<Entry<T>>,
+    reply?: Reply | undefined,
+  ) {
+    if (reply) {
+      for (const requestor of this.#requestors) {
+        requestor.newData(version, data, reply);
+      }
+    } else {
+      for (const listener of this.#downstreams) {
+        listener.newData(version, data, reply);
+      }
+    }
+  }
+
+  messageUpstream(message: Request, downstream: Listener<T>): void {
+    this.#requestors.add(downstream);
+    this.#upstream?.messageUpstream(message);
+  }
+
+  commit(version: Version) {
+    if (this.#requestors.size > 0) {
+      this.#requestors;
+      for (const requestor of this.#requestors) {
+        try {
+          requestor.commit(version);
+        } catch (e) {
+          this.#requestors.clear();
```

Comment:
can you explain this?

Reply:
`commit` will call client code. If client code throws I think it'd be best to put IVM back into a consistent state by clearing this then re-throwing.